In [1]:
print("Name: Pooja P Rawool")
print("ERP ID: 1132200081")

Name: Pooja P Rawool
ERP ID: 1132200081


In [2]:
import pandas as pd  
import numpy as np  
import matplotlib.pyplot as plt  
from sklearn.preprocessing import LabelEncoder

In [3]:
%cd D:\MSC\Trimester3\ML2\Datasets

D:\MSC\Trimester3\ML2\Datasets


# Wilt (Training)

In [41]:
data=pd.read_csv('training.csv')

In [42]:
data.shape

(4339, 6)

In [43]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4339 entries, 0 to 4338
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   class       4339 non-null   object 
 1   GLCM_pan    4339 non-null   float64
 2   Mean_Green  4339 non-null   float64
 3   Mean_Red    4339 non-null   float64
 4   Mean_NIR    4339 non-null   float64
 5   SD_pan      4339 non-null   float64
dtypes: float64(5), object(1)
memory usage: 203.5+ KB


In [44]:
data.describe()

GLCM_pan   Mean_Green     Mean_Red     Mean_NIR       SD_pan
count  4339.000000  4339.000000  4339.000000  4339.000000  4339.000000
mean    126.831298   233.906908   117.292439   534.104683    24.924588
std      13.735836    60.757687    60.711159   154.495500    11.008303
min       0.000000   164.625000    59.142857    86.500000     0.000000
25%     118.589080   206.000000    91.975244   422.875000    18.009143
50%     127.479167   221.454545   101.727273   528.500000    23.612444
75%     135.043591   241.791304   116.866071   643.087037    29.899148
max     183.281250   955.714286   746.333333  1005.516129   156.508431

In [45]:
data.head()

class    GLCM_pan  Mean_Green    Mean_Red    Mean_NIR     SD_pan
0     w  120.362774  205.500000  119.395349  416.581395  20.676318
1     w  124.739583  202.800000  115.333333  354.333333  16.707151
2     w  134.691964  199.285714  116.857143  477.857143  22.496712
3     w  127.946309  178.368421   92.368421  278.473684  14.977453
4     w  135.431548  197.000000  112.690476  532.952381  17.604193

In [46]:
data["class"].value_counts()

n    4265
w      74
Name: class, dtype: int64

2% of the tuples are having class w and the remaining 98% are having the class n

In [47]:
y=data["class"]
x=data.drop("class",axis=1)

In [48]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [49]:
x_test.shape

(868, 5)

In [50]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(solver='liblinear',C=10,random_state=0).fit(x_train,y_train)

In [13]:
y_pred = clf.predict(x_test)

In [14]:
clf.score(x_test, y_test)

0.9815668202764977

Unfortunately, the accuracy is misleading for the tuples with class "w" I'm getting 98% of the accuracy.

This is clearly a problem because many machine learning algorithms are designed to maximize overall accuracy.

In [15]:
from sklearn.metrics import confusion_matrix
conf_mat = confusion_matrix(y_test,y_pred)
print(conf_mat)

[[848   4]
 [ 12   4]]


12 tuples were having class "w" as its actual class and were missclassified as class "n" and 4 tuples were having class "n" as its actual class and missclassified as class "w"

Which means more tuples with class "w" were missclassified.

As tuples with class "w" are less i.e. 74. I am considering class "w" as the class of my interest

In [16]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           n       0.99      1.00      0.99       852
           w       0.50      0.25      0.33        16

    accuracy                           0.98       868
   macro avg       0.74      0.62      0.66       868
weighted avg       0.98      0.98      0.98       868



Precision:<br>
out of all tuples with class "n" prediction 99% I got it right.<br>
out of all tuples with class "w" prediction 50% I got it right.<br>
Recall:<br>
out of all tuples with class "n" truth 100% I got it right.<br>
out of all tuples with class "w" truth 25% I got it right.<br>

In [17]:
from sklearn.metrics import roc_auc_score
prob=clf.predict_proba(x_test)

In [18]:
prob

array([[9.98568706e-01, 1.43129445e-03],
       [9.95702901e-01, 4.29709920e-03],
       [9.99978164e-01, 2.18360409e-05],
       ...,
       [9.97436076e-01, 2.56392398e-03],
       [9.97561514e-01, 2.43848625e-03],
       [9.99998642e-01, 1.35784765e-06]])

In [19]:
# keep only positive class
prob=[p[1] for p in prob]

In [20]:
prob

[0.001431294450063157,
 0.004297099197473105,
 2.1836040935273626e-05,
 0.00039815921401082044,
 0.0005022083300219517,
 8.210656451100475e-08,
 8.161009690378202e-05,
 0.018045986981029803,
 0.0010534765010115528,
 0.2801000942832516,
 0.001336629898747053,
 2.385193305341403e-11,
 0.001498264094184977,
 0.051876289109063084,
 0.0002753106852580246,
 0.0002862021363888146,
 1.0351345902215144e-08,
 0.00025448015662118934,
 0.001201291672383867,
 0.00019288468053240554,
 8.027379579761458e-05,
 0.00046983644785404503,
 0.017455600632728482,
 0.00033062333256369045,
 6.201095820720475e-07,
 8.346160152981752e-06,
 0.0003942082458476255,
 2.1878356818909593e-05,
 0.004691536522127213,
 0.000127068384876538,
 5.383620617264982e-08,
 0.0008637107492209215,
 0.0005355252685694108,
 0.0004017822975425298,
 0.0004131928900912606,
 0.00015223192404479046,
 0.00036800175465976384,
 4.135492166595218e-05,
 1.880504454464069e-05,
 2.518633760257883e-05,
 0.004933541457594272,
 0.00215483945509991

In [21]:
print(roc_auc_score(y_test,prob))

0.9878961267605634


# Random under sampling

### Separate majority and minority classes

In [51]:
data["class"].value_counts()

n    4265
w      74
Name: class, dtype: int64

In [52]:
data_majority=data[data["class"]=="n"]
data_minority=data[data["class"]=="w"]

### Downsample majority class

In [53]:
from sklearn.utils import resample
data_majority_downsampled=resample(data_majority,replace=False,n_samples=74,random_state=123)

### Combine minority class with downsampled majority class

In [54]:
data_downsampled=pd.concat([data_majority_downsampled,data_minority])

### Display new class count

In [55]:
data_downsampled["class"].value_counts()

w    74
n    74
Name: class, dtype: int64

In [56]:
data_downsampled.shape

(148, 6)

As you can see new dataframe has fewer observations than the original and the ratio of 2 classes is now 1:1.

In [57]:
y=data_downsampled["class"]
x=data_downsampled.drop("class",axis=1)

In [58]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [59]:
x_test.shape

(30, 5)

In [60]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(solver='liblinear',C=10,random_state=0).fit(x_train,y_train)

In [61]:
y_pred = clf.predict(x_test)

In [62]:
clf.score(x_test, y_test)

0.9666666666666667

With down sampling accuracy has reduced to 96%

In [63]:
from sklearn.metrics import confusion_matrix
conf_mat = confusion_matrix(y_test,y_pred)
print(conf_mat)

[[16  0]
 [ 1 13]]


1 tuple was having class "n" as its actual class and was missclassified as class "w" 

Only one tuple with class "n" were missclassified.

As tuples with class "w" are less i.e. 74. I am considering class "w" as the class of my interest. Random under sampling is giving good results.

In [34]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           n       0.94      1.00      0.97        16
           w       1.00      0.93      0.96        14

    accuracy                           0.97        30
   macro avg       0.97      0.96      0.97        30
weighted avg       0.97      0.97      0.97        30



Precision:<br>
out of all tuples with class "n" prediction 94% I got it right.<br>
out of all tuples with class "w" prediction 100% I got it right.<br>
Recall:<br>
out of all tuples with class "n" truth 100% I got it right.<br>
out of all tuples with class "w" truth 93% I got it right.<br>

In [35]:
from sklearn.metrics import roc_auc_score
prob=clf.predict_proba(x_test)

In [36]:
prob

array([[9.99921580e-01, 7.84202761e-05],
       [9.99999518e-01, 4.82406567e-07],
       [1.00000000e+00, 1.48946452e-18],
       [1.01818436e-04, 9.99898182e-01],
       [1.00000000e+00, 2.82630464e-10],
       [9.95459182e-01, 4.54081760e-03],
       [9.99881022e-01, 1.18977640e-04],
       [9.99789620e-01, 2.10379590e-04],
       [3.80174868e-06, 9.99996198e-01],
       [9.99769558e-01, 2.30441810e-04],
       [1.30427945e-04, 9.99869572e-01],
       [9.99999279e-01, 7.20779440e-07],
       [1.14778697e-02, 9.88522130e-01],
       [1.55120660e-06, 9.99998449e-01],
       [1.40353533e-02, 9.85964647e-01],
       [9.95933120e-01, 4.06687951e-03],
       [9.93653111e-01, 6.34688877e-03],
       [9.62802487e-01, 3.71975127e-02],
       [9.99996230e-01, 3.77002751e-06],
       [9.99979743e-01, 2.02573345e-05],
       [1.00000000e+00, 5.52415887e-21],
       [3.93768646e-05, 9.99960623e-01],
       [2.07660757e-03, 9.97923392e-01],
       [3.79545683e-03, 9.96204543e-01],
       [1.362896

In [37]:
# keep only positive class
prob=[p[1] for p in prob]

In [38]:
prob

[7.842027610624633e-05,
 4.82406566778137e-07,
 1.489464519360008e-18,
 0.9998981815639711,
 2.826304635251344e-10,
 0.004540817597378718,
 0.00011897764038487857,
 0.00021037958970553181,
 0.9999961982513195,
 0.00023044181041192345,
 0.9998695720550538,
 7.207794397239622e-07,
 0.9885221302503242,
 0.9999984487933993,
 0.9859646466818904,
 0.0040668795105310305,
 0.006346888765800946,
 0.0371975127335592,
 3.7700275055587343e-06,
 2.0257334528175217e-05,
 5.524158869597424e-21,
 0.9999606231353988,
 0.9979233924325904,
 0.9962045431727994,
 0.8637103076566139,
 3.020510221621094e-07,
 0.9846484661403915,
 0.9880383417266274,
 0.2551134877181252,
 0.8758509392353003]

In [39]:
print(roc_auc_score(y_test,prob))

1.0


With respect to the original model trained on the imbalanced dataset got the AUROC score of 98% and when the dataset was down sampled got the AUROC score of 100%.

# Random over sampling

### Upsample minority class

In [65]:
from sklearn.utils import resample
data_minority_upsampled=resample(data_minority,replace=True,n_samples=4265,random_state=123)

### Combine majority class with upsampled minority class

In [66]:
data_upsampled=pd.concat([data_minority_upsampled,data_majority])

In [67]:
data_upsampled.shape

(8530, 6)

As you can see new dataframe has more observations than the original and the ratio of 2 classes is now 1:1.

### Display new class count

In [68]:
data_upsampled["class"].value_counts()

w    4265
n    4265
Name: class, dtype: int64

In [69]:
y=data_upsampled["class"]
x=data_upsampled.drop("class",axis=1)

In [70]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [71]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(solver='liblinear',C=10,random_state=0).fit(x_train,y_train)

In [72]:
y_pred = clf.predict(x_test)

In [73]:
clf.score(x_test, y_test)

0.9859320046893317

With up sampling the accuracy seems to be higher as compared to the down sampling accuracy.

In [74]:
from sklearn.metrics import confusion_matrix
conf_mat = confusion_matrix(y_test,y_pred)
print(conf_mat)

[[826  24]
 [  0 856]]


24 tuples were having class "w" as its actual class and missclassified as class "n".

Which means all the tuples with class "n" were correctly classified. 

In [24]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           n       1.00      0.97      0.99       850
           w       0.97      1.00      0.99       856

    accuracy                           0.99      1706
   macro avg       0.99      0.99      0.99      1706
weighted avg       0.99      0.99      0.99      1706



Precision:<br>
out of all tuples with class "n" prediction 100% I got it right.<br>
out of all tuples with class "w" prediction 97% I got it right.<br>
Recall:<br>
out of all tuples with class "n" truth 97% I got it right.<br>
out of all tuples with class "w" truth 100% I got it right.<br>

In [25]:
from sklearn.metrics import roc_auc_score
prob=clf.predict_proba(x_test)

In [26]:
prob

array([[3.55004845e-02, 9.64499515e-01],
       [4.15668107e-02, 9.58433189e-01],
       [1.00000000e+00, 9.74834681e-12],
       ...,
       [2.45256391e-01, 7.54743609e-01],
       [3.00686352e-03, 9.96993136e-01],
       [9.51486094e-03, 9.90485139e-01]])

In [27]:
# keep only positive class
prob=[p[1] for p in prob]

In [28]:
prob

[0.9644995154902734,
 0.9584331893091543,
 9.748346805265548e-12,
 0.9089050588374346,
 0.0004558252738637044,
 0.005243308730040504,
 0.997927779238857,
 0.9981583777902324,
 0.005345461367799481,
 0.9870820110050345,
 0.29921686067005576,
 0.005502348452103933,
 4.2830650298028055e-42,
 0.05164213696719961,
 0.9993263756045947,
 0.9983082474407077,
 0.9785146651514285,
 0.01359153963966069,
 0.040455758221494935,
 0.0002989405708403681,
 0.0006862869495431916,
 0.9148151801700621,
 2.6902748637887093e-05,
 0.9987036683452534,
 0.9969931364795882,
 0.6099593047968868,
 0.9968605792667944,
 0.9089050588374346,
 0.9809778505433208,
 0.00013650001789519892,
 2.2786090778916727e-05,
 0.9615333784411136,
 0.9987036683452534,
 0.9319609599581717,
 0.9621331874245285,
 0.1946495793829716,
 0.014683183136849425,
 3.528936207770333e-05,
 0.9987036683452534,
 0.0015454647050944569,
 0.908204743985857,
 3.4873621595809645e-05,
 0.04867829942660536,
 0.9821809220519538,
 0.989193552760591,
 2.331

In [29]:
print(roc_auc_score(y_test,prob))

0.9933617372182517


With respect to the original model trained on the imbalanced dataset got the AUROC score of 98% and when the dataset was up sampled got the AUROC score of 99%.

# SMOTE

In [90]:
y=data["class"]
x=data.drop("class",axis=1)

In [91]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [92]:
from imblearn.over_sampling import SMOTE
oversample = SMOTE()
x_train_smote, y_train_smote = oversample.fit_resample(x_train, y_train)

In [93]:
x_train_smote1=x_train_smote.copy()
y_train_smote1=y_train_smote.copy()

In [94]:
df_smote=x_train_smote1

In [95]:
df_smote["class"]=y_train_smote1

In [96]:
df_smote["class"].value_counts()

w    3413
n    3413
Name: class, dtype: int64

In [97]:
df_smote.shape

(6826, 6)

In [98]:
x_train.shape

(3471, 5)

In [99]:
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(solver='liblinear',C=10,random_state=0).fit(x_train_smote,y_train_smote)

In [100]:
y_pred = clf.predict(x_test)

In [101]:
clf.score(x_test, y_test)

0.9769585253456221

In [102]:
from sklearn.metrics import confusion_matrix
conf_mat = confusion_matrix(y_test,y_pred)
print(conf_mat)

[[833  19]
 [  1  15]]


1 tuple was having class "n" as its actual class and were missclassified as class "w" and 19 tuples were having class "w" as its actual class and missclassified as class "n".

tuples with class "w" which were missclassified are more.

In [103]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           n       1.00      0.98      0.99       852
           w       0.44      0.94      0.60        16

    accuracy                           0.98       868
   macro avg       0.72      0.96      0.79       868
weighted avg       0.99      0.98      0.98       868



Precision:<br>
out of all tuples with class "n" prediction 100% I got it right.<br>
out of all tuples with class "w" prediction 44% I got it right.<br>
Recall:<br>
out of all tuples with class "n" truth 98% I got it right.<br>
out of all tuples with class "w" truth 94% I got it right.<br>

In [104]:
from sklearn.metrics import roc_auc_score
prob=clf.predict_proba(x_test)

In [105]:
prob

array([[9.99297636e-01, 7.02363938e-04],
       [9.91098697e-01, 8.90130308e-03],
       [9.99987821e-01, 1.21787525e-05],
       ...,
       [9.96453233e-01, 3.54676675e-03],
       [9.98485838e-01, 1.51416190e-03],
       [9.99999913e-01, 8.67571162e-08]])

In [106]:
# keep only positive class
prob=[p[1] for p in prob]

In [107]:
prob

[0.0007023639375827317,
 0.00890130307779353,
 1.2178752522496412e-05,
 0.0005170311101261804,
 0.00021595687420998864,
 5.113125179290234e-10,
 1.0946372162506357e-05,
 0.03751699853993051,
 0.002449746055965897,
 0.9370176980293764,
 0.0008950546795059845,
 9.979146557770558e-16,
 0.002526272396857638,
 0.39125391531517595,
 5.709551121061316e-05,
 0.00013744851892045514,
 7.313644326787584e-12,
 9.035103173430136e-05,
 0.0014999870321096489,
 9.219590134505361e-05,
 2.7308180493991847e-05,
 0.00025285795070664405,
 0.029067188095184632,
 0.00015647624249868008,
 6.224972152998972e-09,
 8.052588573880881e-07,
 0.0005223165526712633,
 1.4366492755854091e-05,
 0.010431146715840256,
 5.324137422155663e-05,
 2.3078185869209207e-10,
 0.0007866765841513467,
 0.00025307176195953795,
 0.0010010513831864383,
 0.00018128007389878924,
 4.898395686389944e-05,
 7.711076558177443e-05,
 1.1496364328217643e-05,
 2.893703159626907e-06,
 5.0864802169331675e-06,
 0.023851989919867582,
 0.00376267569585

In [108]:
print(roc_auc_score(y_test,prob))

0.9897300469483569


With respect to the original model trained on the imbalanced dataset got the AUROC score of 98.78% and after applying SMOTE on the dataset got the AUROC score of 98.97%.

Hence no sampling method is guaranteed to work in all problem domains and different solutions must be tried.